Large-scale load testing article (#15)
loadimpact/k6-docs: src/data/markdown/docs/01 guides/05 Testing Guides/04 Large-scale k6 testing.md

For monitoring network, we recommend [iftop](https://en.wikipedia.org/wiki/Iftop)

Here's a screenshot of 3 terminal sessions showing k6, iftop and htop.
![k6 iftop htop](./images/large-scale-testing-3-terminals.png)

## Additional k6 flags to achieve better performance

### --compatibility-mode=base

If you are pushing the limits of the hardware, this is the most impactful k6 setting you can enable.

This setting disables the internal babel transpilation from ES6 to ES5 and inclusion of corejs library.

<div class="doc-blockquote" data-props='{}'>

> ### Some background
> k6 at its core executes ECMAScript 5.1 code. Most k6 script examples and documentation is written in ECMAScript 6+. 
> When k6 gets ES6+ code as input, it transpiles it to ES5.1 using babel and loads corejs to enable commonly used APIs.
> This works very well for 99% of use cases, but it adds significant overheard with large tests.

</div>

To get the best performance out of k6, it's best to transpile the scripts outside of k6 using webpack.

In this repository we have prepared an efficient transpilation scheme that produces performant ES5.1 code for k6. 

Use it like this:

<div class="code-group">

```bash

> git clone https://github.com/loadimpact/k6-hardware-benchmark/

Author: meh, I like copy-pasting myself ^^

---

> title: "Hardware-sizing for large-scale k6 tests"

Reviewer: Maybe "Hardware requirements for large-scale k6 tests" or "Determining hardware requirements ..."?

Or "resource requirements" considering the article also discusses OS tweaks?

Reviewer 2: I vote for hardware requirements

Author: Change per your recommendation :+1:

excerpt: "How to run large-scale k6 tests without distributed-execution"
---

This document explains how to launch a large-scale k6 test on a single machine without the need of distributed execution. 

> The common misconception of many load-testers is that distributed-execution (ability to launch a load test on multiple machines) is required to generate large load. This is not the case with k6.

Author: Not sure why I put hyphens everywhere :D

k6 is different from many other load-testing tools in a way it handles hardware resources. A single k6 process will efficiently use all CPU cores on a load-generator machine.

> Single instance of k6 is often enough to generate load of 30.000-40.000 simultaneus users (VUs). This amount of VUs can generate upwards of 300.000 requests per second (RPS). 

Reviewer: ```suggestion
A single instance of k6 is often enough to generate load of 30.000-40.000 simultaneous users (VUs). This amount of VUs can generate upwards of 300.000 requests per second (RPS). 
```

Unless you need more than 100.000-300.000 requests per second (6-12M requests per minute), a single instance of k6 will likely be sufficient for your needs.

Below we will explore what hardware is needed for generating different levels of load.

<div class="doc-blockquote">

> ### Note about native distributed-execution in k6
>
> The long-term goal for k6 is to support distributed execution natively. We are currently laying groundwork for this feature in [PR #1007](https://github.com/loadimpact/k6/pull/1007).

> > You can follow this effort on guthub [in issue 140](https://github.com/loadimpact/k6/issues/140). Again, this is useful only if you need to generate load larger than 300k RPS.

Reviewer: ```suggestion
> You can follow this effort on GitHub [in issue 140](https://github.com/loadimpact/k6/issues/140). Again, this is useful only if you need to generate load larger than 300k RPS.
```
:smile:

Reviewer 2: I think we should move to guthub, sounds awesome!

>
> [k6 cloud](/cloud) (a paid service) supports distributed execution already.

</div>

## OS fine-tuning for maximum performance

> For the purpose of this demonstration, we are using a Linix (Ubuntu Server) machine. The instructions will be the same for any Linux distribution. 

Reviewer: ```suggestion
For the purpose of this demonstration, we are using a Linux (Ubuntu Server) machine. The instructions will be the same for any Linux distribution. 
```

Author: Damn, I should have ran a spell checker before committing :D

When running the test, you can use `iftop` in the terminal to view in real time the amount of network traffic generated.
If the traffic is constant at 1Gbit/s, your test is probably limited by the network card. Consider upgrading to a different EC2 instance.

### CPU
Unlike many other load testing tools, k6 is heavily multithreaded. It will effectively use all available CPU cores.

The amount of CPU you need depends on your test files (sometimes called test script). 
Regardless of the test file, you can assume that large tests require significant amount of CPU power. 
We recommend that you size the machine to have at least 20% idle cycles (up to 80% used by k6, 20% idle). 
If k6 uses 100% to generate load, it won't have enough CPU to measure the responses correctly. 
This may cause the result metrics to have much larger response time than in reality.

### Memory
k6 likes memory, but [it isn't as greedy as other load testing tools](https://k6.io/blog/comparing-best-open-source-load-testing-tools#memory-usage). 
Memory consumption heavily depend on your test scenarios. To estimate the memory requirement of your test, 

> run the test on your laptop with 100VUs and multiply the consumed memory by the target number of VUs. 

Reviewer 3: ```suggestion
run the test on your development machine with 100VUs and multiply the consumed memory by the target number of VUs. 
```
I think "work machine" is too wide of a definition .. but not all people work/develop on laptops ;)

});
```

</div>

Code like this runs fine when the system under test (SUT) is not overloaded and returns proper responses.
When the system starts to fail, the above check won't work as expected. 

The issue here is that the check assumes that there's  always a body in a response. The `r.body` may not exist if server is failing.
In such case the check itself won't work as expected and error similar to the one below will be returned: 

<div class="code-group" data-props='{}'>

```
ERRO[0625] TypeError: Cannot read property 'length' of undefined
```

</div> 

To fix this issue your checks must be resilient to any response type. This change will fix the above problem.

<div class="code-group" data-props='{"labels": ["resilient check"]}'>

```javascript
let checkRes = check(res, {
    "Homepage body size is 11026 bytes": (r) => r.body && r.body.length === 11026
});
```

</div>

### Monitor the load-generator server

If you are running a test for the first time, it's a good idea to keep an eye on the available resources while the test is running.
The easiest way to do so is to ssh to the server with 3 sessions:
1. To run k6 
2. To monitor CPU and memory

> 3. To monitor the network

Reviewer: Maybe also mention screen/tmux here? ;)

For monitoring CPU and memory we recommend [htop](https://en.wikipedia.org/wiki/Htop). 
For monitoring network, we recommend [iftop](https://en.wikipedia.org/wiki/Iftop)

Here's a screenshot of 3 terminal sessions showing k6, iftop and htop.
![k6 iftop htop](./images/large-scale-testing-3-terminals.png)

## Additional k6 flags to achieve better performance

### --compatibility-mode=base

If you are pushing the limits of the hardware, this is the most impactful k6 setting you can enable.

> This setting disables the internal babel transpilation from ES6 to ES5 and inclusion of corejs library.

Reviewer 3: maybe some links ?
Also it is ES5.1+ ;)

> This setting disables the internal babel transpilation from ES6 to ES5 and inclusion of corejs library.

Reviewer 4: Maybe capital B on "babel" and link to https://babeljs.io/?

Reviewer 4: Same with "corejs"

<div class="doc-blockquote" data-props='{}'>

> ### Some background
> k6 at its core executes ECMAScript 5.1 code. Most k6 script examples and documentation is written in ECMAScript 6+. 

> > When k6 gets ES6+ code as input, it transpiles it to ES5.1 using babel and loads corejs to enable commonly used APIs.

Reviewer 3: it ALWAYS loads corejs even if babel wasn't used

> This works very well for 99% of use cases, but it adds significant overheard with large tests.

</div>

To get the best performance out of k6, it's best to transpile the scripts outside of k6 using webpack.

In [k6-hardware-benchmark](https://github.com/loadimpact/k6-hardware-benchmark) repository, we have prepared an efficient transpilation scheme that produces performant ES5.1 code for k6. 

Use it like this:

<div class="code-group">

```bash
git clone https://github.com/loadimpact/k6-hardware-benchmark/
cd k6-hardware-benchmark
yarn install
yarn run to-es5 someplace/yourscript.js
# your ES5 script is in someplace/yourscript.es5.js
```
</div>

Once your code is transpiled, run it like this:

<div class="code-group">

```bash
k6 run -o cloud --compatibility-mode=base someplace/yourscript.es5.js
```
</div>

> k6 will use about 50% of memory in comparison to running the original script. It will also reduce the CPU load. 

Reviewer 3: I would argue that if your script is simple and uses discardResponseBodies the difference is more like [85%](https://github.com/loadimpact/k6/issues/1167#issuecomment-553835092) , with a 1.5mb json data file(If my memory served) the difference was 50%, but I can't currently find where I've said that :sob: . I would argue some links to benchmark data will be useful here.

Also another thing that compatibility-mode=base does is "significantly" increase the startup time. In the above linked script I ran the script for 60s and in the one case there were additiona minute and 5 seconds (on average) additiona time in the other a secodn and a half for 3500 VUs. Which IMO is "significant" :D

Author: Updated the text. If you have relevant benchmark links, let me know.

### discardResponseBodies

You can tell k6 to not process the body of the response by setting `discardResponseBodies` in the options object like this:

<div class="code-group">

```js
export let options = {
	discardResponseBodies: true,
};
```
</div>

> k6 by default loads the response body of the request into memory. This causes much higher memory consumption, and often is completely unnecessary.

Reviewer 3: You can add "but if you need it for some requests you can use [Params.responseType](https://k6.io/docs/javascript-api/k6-http/params-k6-http)" , but with an example

### --no-thresholds --no-summary 

> If you are running a cloud test with local execution (`k6 run -o cloud`), you may want to disable the terminal summary 

Reviewer 4: "running a cloud test with local execution" sounds strange to me, I'd phrase it as "running a local test and streaming results to the cloud"

If you have configured [abortOnFail thresholds](https://k6.io/docs/using-k6/thresholds#aborting-a-test-when-a-threshold-is-crossed), k6 needs to evaluate the result constantly to verify that the threshold wasn't crossed. Consider removing this setting.

## File upload testing
Special considerations must be taken when testing file-uploads. 

### Network throughput
The network throughput of the load-generator machine, as well as the SUT will likely be the bottleneck. 

### Memory
k6 needs significant amount of memory when uploading files, as every VU is independent and has its own memory.

### Data transfer costs
k6 can upload a large amount data in a very short period of time. Make sure you understand the data transfer costs before commencing a large scale test.

[Outbound Data Transfer is expensive in AWS EC2](https://www.cloudmanagementinsider.com/data-transfer-costs-everything-you-need-to-know/). The price ranges between $0.08 to $0.20 per GB depending on the region. 
If you use the cheapest region the cost is about $0.08 per GB. Uploading 1TB therefore costs about $80. Long running test can cost several hundreds of dollars in data transfer alone.

### EC2 costs

> The AWS EC2 instances are relatively cheap. Even the largest instance we have used in this benchmark (m5.24xlarge) costs only $4.6 per hour. Make sure to turn off the load-gen servers once you are done with your testing. Forgotten EC2 server will cost $3312 per month.  

Reviewer 4: "load-gen" -> "load generator" for consistency?

Tip: it's often possible to launch "spot instances" of the same hardware for 10-20% of the cost. 

> ## Errors

Reviewer 4: This section is really good, but it feels like it shouldn't be in its own article on "Understanding k6 errors" or something like that.

Reviewer 2: @robingustafsson I suspect there's a negation too much in that sentence? 😅  If so, I agree that it would be nice to have in its own article.

Author: :+1: I'll release this as is, and then branch it out to another article :-)

```
WARN[0064] Request Failed    error="Get http://test.k6.io: context deadline exceeded"
```

</div>

### dial tcp 52.18.24.222:80: i/o timeout

This is a similar error to the one above, but in this case k6 wasn't even able to make a request. The target system isn't able to establish a connection.

<div class="code-group">

```
WARN[0057] Request Failed     error="Get http://pawel.staging.loadimpact.com/static/logo.svg?url=v3: dial tcp 52.18.24.222:80: i/o timeout"
```

</div>

### socket: too many open files

This error means that the load-generator machine isn't able to open TCP sockets because it reached the limit of open file descriptors.
Make sure that your limit is set sufficiently high `ulimit -n 250000` should be enough for anyone :tm:

<div class="code-group">

```
WARN[0034] Request Failed     error="Get http://99.81.83.131/static/logo.svg?ip=6: dial tcp 99.81.83.131:80: socket: too many open files"
```

</div>

Note: you should decide what level of errors is acceptable. At large scale some errors are always present.
If you make 50M requests with 100 failures, this is generally a good result (0.00002% errors).

# Benchmarking k6 on AWS hardware

> We have executed few large tests on different EC2 machines to see how much load k6 can generate. 

Reviewer 4: "few" -> "a few"

Our general observation is that k6 scales proportionally to the hardware. 2x larger machine is able to generate 2x more traffic.

> The limit to this scalability is in the number of open connections. Single Linux machine can open up to `65 535` sockets per IP. 

Reviewer: ```suggestion
The limit to this scalability is in the number of open connections. A single Linux machine can open up to `65 535` sockets per IP. 
```

> The limit to this scalability is in the number of open connections. Single Linux machine can open up to `65 535` sockets per IP. 

Reviewer 4: nitpick but it's per <src IP, src port, dst IP, dst port> tuple afaik, so should be able to go beyond 65k if machine has multiple nics with different IPs or target system hostnam is behind a load balancer with different IPs etc.

Author: I'm happy to update that, but I think I'll have to mention that k6 doesn't have support for multiple IPs and discuss this for a while. I don't want to go too much off-topic :-) (here's another hyphen!)

I also think that the limit of 65k is for <src IP, src port>. I don't think the destination IP matters when all 65k ports are already used by the load-gen server. (or maybe I'm missing something).

This means that maximum of 65k requests can be executed simultaneously on a single machine.
The RPS limit depends on the response time of the SUT. If responses are delivered in 100ms, the RPS limit is 650 000. 

## Real-life test of a website.

Testing the theoretical limits is fun, but that's not the point of this benchmark. 
The point of this benchmark is to give users indication of how much traffic k6 can generate when executing complicated, real-life tests.
For this purpose we have written a rather heavy [real-life website test](https://github.com/loadimpact/k6-hardware-benchmark/blob/master/scripts/website.js) that uses almost all k6 features.

Setup:
- All tests were executed on AWS EC2 instances

> - The "discardResponseBodies" recommendation was NOT used. (results would be better with this setting).

Reviewer 3: I would argue you can try to do a test with this and see how much better they will be :). This way when someone doesn't use it  and looks at optimizing they can see some example of what the difference si

Author: yeah, that's a fair point, but I don't want to go back to testing for this article :-) 
I'll maintain this page, and will likely update it later.

- Note: the target system (test.k6.io) is a slow-ish PHP website, not optimized for performance - a static website would be much quicker.

The "website.js" test file uses a wide range of k6 features to make the test emulate a real usage of k6. This is not a test rigged for performance - quite the opposite.
This test uses plenty of custom metrics, checks, parametrization, batches, thresholds and groups. It's a heavy test that should represent well the "real life" use case.

### Execution on AWS m5.large EC2 server

The `m5.large` instance has 8GB of RAM and 2 CPU cores. 

The following command was used to execute the test
<div class="code-group">

```bash
k6 run scripts/website.es5.js \
 -o cloud \
 --vus=6000 \
 --duration=10m \
 --compatibility-mode=base \
 --no-thresholds \
 --no-summary
```

</div>

Results
- Maximum VUS reached: 6000
- Memory used: 6.09 GB  (out of 8.0)
- CPU load (avg): 1.49 (out of 2.0). 
- Peak RPS: ~6000 (note, this test was not optimized for RPS).
- 2x `sleep(5)` in each iteration.

### Execution on AWS m5.4xlarge
The `m5.4xlarge` instance has 64GB of RAM and 16 CPU cores.

> https://app.k6.io/runs/720179

Reviewer 3: Is this going to be a public link? Will the other ones have public links as well ?

Author: Ah, missed it. I'll remove these links.

<div class="code-group">

```bash
k6 run scripts/website.es5.js \
   -o cloud  \
   --vus=20000 \ 
   --duration=10m \ 
   --compatibility-mode=base  \
   --no-thresholds  \
   --no-summary 

```

</div>

Results
- Maximum VUS reached: 20.000
- Memory used: 20.1 GB  (out of 61.4)
- CPU load (avg): 8.5 (out of 16.0). 
- Peak RPS: ~20.000 (note, this test was not optimized for RPS).
- 2x `sleep(5)` in each iteration.

### Execution on AWS m5.24xlarge
The m5.24xlarge has 384GB of RAM and 96 CPU cores.
NOTE: sleep has been reduced to 1s instead of 5s to produce more requests.
<div class="code-group">

```bash
k6 run scripts/website.es5.js  \
   -o cloud  \
   --vus=30000 \ 
   --duration=5m \ 
   --compatibility-mode=base  \
   --no-thresholds  \
   --no-summary 
```

</div>

Results
- Maximum VUS reached: 30.000
- Memory used: ~120 GB  (out of 370 available)
- CPU load (avg): ~45 (out of 96.0). 
- Peak RPS: ~61.500.
- `sleep(1)` in each iteration.

> ## Testing for RPS.

Reviewer 4: No other heading has a dot at the end 🙂

Please read the warning about the cost of data transfer in AWS before commencing a large scale test.

### Execution on AWS m5.24xlarge

To test the network throughput we have written a [file-uploading script](https://github.com/loadimpact/k6-hardware-benchmark/blob/master/scripts/file-upload.js). We have executed this test for only 1 minute to minimize the data-transfer costs. In 1 minute, k6 managed to transfer 36 GB of data with 1000 VUs. 

<div class="code-group">

```bash
k6 run scripts/file-upload.es5.js \
-o cloud \
--vus=1000 \ 
--duration=1m \ 
--compatibility-mode=base \
--no-thresholds \
--no-summary
```

</div>

Results
- Maximum VUS reached: 1.000
- Memory used: 81 GB  (out of 370 available)
- CPU load (avg): 9 (out of 96.0). 
- Network throughput reached **4.7Gbit/s**
- Data transferred: 36GB.

Note: each VU in k6 is completely independent, and therefore it doesn't share any memory with other VUs. 
1000VUs uploading 26MB file need as much as 81GB of RAM since each VU holds the copy of the file in memory. 

## Summary

k6 is able to fully utilize CPU, memory and Network bandwidth available on any hardware we have tested it on. 

> Single instance of k6 can run 30k+ VUs and produce 100k+ RPS. For the vast majority of systems, load coming from a single k6 process will be more than enough.

Reviewer: ```suggestion
A single instance of k6 can run 30k+ VUs and produce 100k+ RPS. For the vast majority of systems, load coming from a single k6 process will be more than enough.
```
